File: 0453-Minimum-Number-of-Arrows-to-Burst-Balloons/0453-Minimum-Number-of-Arrows-to-Burst-Balloons.py
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        ans = 1
        endpoint = float("inf")
        points.sort()
        for start, end in points:
            if endpoint < start:
                ans += 1 
                endpoint = end
            else:
                endpoint = min(endpoint, end)
        return ans
      
# TC : O(nlogn), where n is the number of balloons. ( O(nlogn) + O(n) ): sorting function and for iteration through sorted list of balloons
# SC : O(1), since the solution does not use any additional space beyond the input array.

File: 0453-Minimum-Number-of-Arrows-to-Burst-Balloons/0453-Minimum-Number-of-Arrows-to-Burst-Balloons.cpp
class Solution {
 public:
  int findMinArrowShots(vector<vector<int>>& points) {
    int ans = 1;
    int endpoint = INT_MAX;
    sort(points.begin(), points.end());
    for (const auto& point : points) {
      int start = point[0];
      int end = point[1];
      if (endpoint < start) {
        ans++;
        endpoint = end;
      } 
      else {
        endpoint = min(endpoint, end);
      }
    }
    return ans;
  }
};
